homeassistant/components/text/__init__.py

```diff
@@ -0,0 +1,216 @@
+
+
+async def _async_set_value(entity: TextEntity, service_call: ServiceCall) -> None:
+    """Service call wrapper to set a new value."""
+    value = service_call.data[ATTR_VALUE]
+    if len(value) < entity.min:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} is too short (minimum length {entity.min})"
+        )
+    if len(value) > entity.max:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} is too long (maximum length {entity.max})"
+        )
+    if entity.pattern is not None and re.compile(entity.pattern).match(value) is None:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} doesn't match pattern {entity.pattern}"
+        )
+    await entity.async_set_value(value)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class TextMode(StrEnum):
+    """Modes for text entities."""
+
+    PASSWORD = "password"
+    TEXT = "text"
+
+
+@dataclass
+class TextEntityDescription(EntityDescription):
+    """A class that describes text entities."""
+
+    native_min: int = 0
+    native_max: int = MAX_LENGTH_STATE_STATE
+    mode: TextMode = TextMode.TEXT
+    pattern: str | None = None
+
+
+class TextEntity(Entity):
+    """Representation of a Text entity."""
+
+    entity_description: TextEntityDescription
+    _attr_mode: TextMode
+    _attr_native_value: str | None
+    _attr_native_min: int
+    _attr_native_max: int
+    _attr_pattern: str | None
+    _attr_state: None = None
+
+    @property
+    def capability_attributes(self) -> dict[str, Any]:
+        """Return capability attributes."""
+        return {
+            ATTR_MODE: self.mode,
+            ATTR_MIN: self.min,
+            ATTR_MAX: self.max,
+            ATTR_PATTERN: self.pattern,
+        }
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if self.native_value is None:
+            return None
+        if len(self.native_value) < self.min:
+            raise ValueError(
+                f"Entity {self.entity_id} provides state {self.native_value} which is "
+                f"too short (minimum length {self.min})"
+            )
+        if len(self.native_value) > self.max:
+            raise ValueError(
+                f"Entity {self.entity_id} provides state {self.native_value} which is "
+                f"too long (maximum length {self.max})"
+            )
+        if (
+            self.pattern is not None
+            and re.compile(self.pattern).match(self.native_value) is None
```

Comment:
I think we should avoid compiling it every time, instead, we should compile it once and re-use that pattern.

Not exactly sure how we should do that. Should maybe we require a regex object on the `pattern` property instead of a string?

```py
_attr_pattern: re.Pattern | None
```

Reply:
I made this switch but I think it is better to handle this internally to make it easier for the integration author. I've got an idea on how to do this.